tests/tt_eager/python_api_testing/unit_testing/misc/test_rotary_embedding_llama.py

```diff
@@ -0,0 +1,260 @@
+
+    tt_out = tt_model(*tt_inp)
+    tt_out = [tt2torch_tensor(tt_out_tensor) for tt_out_tensor in tt_out]
+
+    # check outputs ----------------------------------------------------------------------
+    does_pass = True
+    for i in range(2):
+        out_pass, output_pcc = comp_pcc(pytorch_out[i], tt_out[i], pcc)
+        # Check each shape matches
+        assert pytorch_out[i].shape == tt_out[i].shape
+        logger.info(f"PCC value: {output_pcc}")
+        does_pass = does_pass and out_pass
+
+        mae = torch.mean(torch.abs(pytorch_out[i] - tt_out[i]))
+        logger.info(f"MAE: {mae}")
+
+        max_incorrect = torch.max(torch.abs(pytorch_out[i] - tt_out[i]))
+        logger.info(f"Max incorrect: {max_incorrect}")
+
+        max_gt = torch.max(torch.abs(pytorch_out[i]))
+        logger.info(f"Max ground truth: {max_gt}")
+
+    if does_pass:
+        logger.info("Llama QKV output Passed!")
+    else:
+        logger.warning("Llama QKV output Failed!")
+        assert does_pass, f"PCC value is lower than {pcc}"
+
+
+@skip_for_grayskull("Requires eth connected devices to run")
+@pytest.mark.parametrize(
+    "batch, seq_len",
+    (
+        (1, 128),
+        (1, 256),
+        (1, 512),
+        (1, 2048),
+        (1, 4096),
+        (1, 8192),
+        (1, 16384),
+    ),
+    ids=(
+        "prefill_128",
+        "prefill_256",
+        "prefill_512",
+        "prefill_2k",
+        "prefill_4k",
+        "prefill_8k",
+        "prefill_16k",
+    ),
+)
+@pytest.mark.parametrize(
+    "n_heads, n_kv_heads, head_dim",
+    (
+        (8, 1, 64),
+        (8, 1, 128),
+        (11, 3, 128),
+        (71, 32, 64),
+        (8, 1, 96),
+        (8, 1, 256),
+    ),
+)
+@pytest.mark.parametrize("datatype", (ttnn.bfloat16,))
+@pytest.mark.parametrize("pcc", (0.9997,))
+def test_LlamaAttention_inference(
+    batch,
+    seq_len,
+    n_heads,
+    n_kv_heads,
+    head_dim,
+    datatype,
+    pcc,
+    all_devices,
+    use_program_cache,
+):
+    devices = all_devices
+    compute_grid_size = devices[0].compute_with_storage_grid_size()
+    if compute_grid_size.x < 8 or compute_grid_size.y < 8:
+        pytest.skip(f"Requires grid size of at least {(8, 8)} to run")
+
+    max_seq_len = max(4096, seq_len)
+
+    for i in range(3 if use_program_cache else 1):
+        run_test_rotary_embedding_llamma(
+            devices, batch, seq_len, pcc, n_heads, n_kv_heads, head_dim, max_seq_len, datatype
+        )
+
+        # shift input/output tensor by creating very small tensor between loop
+        inp = torch.rand(1, 1, 32, 32)
+        test_tensor = (
+            tt_lib.tensor.Tensor(
+                inp.reshape(-1).tolist(),
+                inp.shape,
+                ttnn.bfloat16,
+                tt_lib.tensor.Layout.ROW_MAJOR,
+            )
+            .to(tt_lib.tensor.Layout.TILE)
+            .to(devices[0])
+        )
+
+    assert devices[0].num_program_cache_entries() == 2
```

Comment:
Please properly test program caching

Reply:
This `test_tensor` doesn't do anything to test cache. It goes out of scope so memory is the same at the start of every loop.

Reply:
appending the `test_tensor` to an array outside of the loop now to make sure it's not going outside of scope.